### DISCUSSION TITLE
"Script" Flow Operation

### DISCUSSION THREAD

**Discussion by claytongulick** [directus/directus#13944] [Feature Requests]
I see how to create a custom extension for Operations in the new Flows functionality, but sometimes this is a bit heavy weight for simple things, like data parsing or transformation that's more complex than what can be handled by the Transform Operation.

Would love to have a new "Script" type operation that would execute arbitrary JavaScript as part of the flow.

The Script Operation should have access to the flow context as well as server-side services that can be used for creating records etc...

It should also have access to environment variables from process.env or the .env file.

Bonus points for being able to require() in external NodeJS modules, though I'm not sure how this would work with node_modules resolution.

Thanks!!

**Comment by benhaynes** [comment]
I love this idea... though it's really a decision for our engineering team, as we need to ensure there are no security issues (especially for our Cloud service).

**Reply by rijkvanzanten** [reply]
Once you get into including node modules, using env vars, and especially API internals, you'd kinda have to start requiring the use of a custom operation extension type instead of something you'd configure through the app 🤔 An app "in-between" solution would be useful for some simpler operations, like simple math, but I think you'd reach a ceiling pretty fast

**Reply by rijkvanzanten** [reply]
(And yes security wise it's not _ideal_ to have arbitrary javascript submitted through the api + saved to the database + executed in a secure environment)

**Reply by SeanDylanGoff** [reply]
**Theoretically** the DX of using an operation extension type should be the same/better because the devs can use a their preferred editor, CI and so on.

In practice i've found that developing extensions is still not as smooth as it could be **in some cases**.
For example npm run dev bundles everything into a single .js - which gets extremely slow with even just a few external dependencies.
I think if the DX for extensions can be improved (further, it's already soooo much better than a few months ago. I've tried and failed to improve it so far, there are a few hard problems there like ES6 imports not having require.cache), the use case for this "in-between" would almost vanish and could be left off. So maybe the time is better spent there?

If we really want an in-app option, maybe it could be an editor that just creates these operations in the extensions folder? And that could be opt-in for security.

**Reply by rijkvanzanten** [reply]
Oh yeah absolutely, extensions DX is gonna be a big focus point in the coming months. That being said, there is still a use case for a "Simple script" operation, as it would allow you to do things like basic math or if-statements without the need to add new code into your stack, in turn requiring  (re)deploys etc. 🤔 It would be sandboxed though, so you could only run "regular javascript" without any module imports or database access or anything like that.

**Comment by colemahatmccreary** [comment]
No idea how relevant to a potential solution, but Zoho has their own JS-like scripting [syntax](https://www.zoho.com/deluge/) for this sort of thing.

Obviously not proper JS, but starting a simple library/parser for Directus might work for those in-between scenarios where offloading to a http/s endpoint (or writing an extension) is too much. 

Plus, starting with super simple operations (math, etc) as mentioned above would allow you to incrementally add to the capabilities over time.

**Reply by rijkvanzanten** [reply]
That'd be a great in-between solution, though I don't think it'll fit for the original post in this thread, as that specifically calls out importing node modules etc as well 🤔

**Reply by benhaynes** [reply]
Would the sandbox be able to use any node modules or libraries we already use with Directus (automatically, or by enabling them with toggles)? We have a lot of useful ones...

**Reply by rijkvanzanten** [reply]
No, none of those, that's where the custom extension type would come in. It's solely meant for things like basic math and other small "inline" operations

**Reply by Attacler** [reply]
> No idea how relevant to a potential solution, but Zoho has their own JS-like scripting [syntax](https://www.zoho.com/deluge/) for this sort of thing.
> 
> Obviously not proper JS, but starting a simple library/parser for Directus might work for those in-between scenarios where offloading to a http/s endpoint (or writing an extension) is too much.
> 
> Plus, starting with super simple operations (math, etc) as mentioned above would allow you to incrementally add to the capabilities over time.

I must say, Zoho Deluge is an example you can use but please keep it as close as possible to the normal Javascript syntax.
Speaking about it, there are some parsers & executors on the web.

However it will always bring security issues with it.
The best way i found (yet) is using https://github.com/laverdet/isolated-vm
This allows you to create a basic JS context and using it you can add more functionality if you need it.
Keep in mind it will cost resources if you plan on using VM's.
(i already made something similar called Dataflows and i'm going to implement them into Operations/flows when the extension type is read)

**Comment by lsanchezs** [comment]
I think is a great idea to have a Script Operation.
I understand it could be generate a potential security issue, my idea would be separate them in two components
**Flows Extensions:** JS Scripts, they must be upload as other extensions (hooks): Not through the App.
**Script Operations:**  DirectusAPP just detects Flows Extensions, listed them, and allow them to linked to Script Operation.

**Reply by henk23** [reply]
I like this approach: Have the custom code inside an extension. No need for it to be uploaded through the app, no need for a sandbox or a VM, no additional security concerns. Just give the app a way of calling the extension code.

**Comment by CHSchuepfer** [comment]
If i may weigh in on that topic as i have basically designed and deployed something like that in the backend of the company i work in.
For the Background i have implemented directus as the middleware with an apache airflow scheduler for a large(ish) scale Selfservice Portal. The only puzzle piece that was missing was to help our Application Managers execute specific scripts without having to give XX people permission on the codebase.

The only solution i came up with that kind of did what i wanted to achieve and keeping up the security was to encapsulate it in an container.
I dont know if that would be a feasible strategy for the your Cloud offering but i had to limit it to 2 Languages only.

I support Python & JS. To Trigger the execution i have a flow that gets triggert when a new "job" is added (so an item.create trigger). Send the Subdata in on the job (Code Textbox) via Rest API to my Airflow Instance. The Airflow dag analyses that script, decided which container to start (python/js) executes it, gets resultset via xcom & excel write back to  directus via graphql call (upload excel & link to item via job id).

To make that make sense in the sense purely for directus i feel like it would need an extension incl. endpoint to do the syntax checking bc if there is no syntax checking you run into trouble. Also you have to be really "on top of it" in terms of packages or 3rd party libraries. Injecting arbitrary 3rd party libs can be a real horror as @rijkvanzanten has mentioned.

So all in all i think its a great idea though, especially i you (directus) hosts a cloud service where one "just kinda checked/parsed" script can have consquences for many others is something that has to be analysed. Like i said i offloaded the biggest part/the hardest part to my airflow instance who does the syntax checking.

So i could see that being extremly usefull (being directly integrated in direcuts) as an extension (maybe after the marketplace launched?) with the express requirement to have an docker engine running (if selfhosted) or in your cloud offering in you CI/CD Pipeline (we use Gitlab).



Just my, very lengthy i know, two cents.


Oh and just bc i dont get tired of saying it: Thank you (the whole directus Team) for your awesome product. I absolutly love the work u do and my project would have gone to sh**s if not for directus (that is the reason why i pledge monthly to you of my own personal money. So yeah thanks & so long

**Reply by benhaynes** [reply]
Thank you for the kind words, and the great direction. Last I spoke with the engineers, the idea of running scripts in a sandbox VM... which sounds like a great path forward.

**Comment by rijkvanzanten** [comment]
Putting this here for later reference: https://thegoodhacker.com/posts/the-unsecure-node-vm-module/ / https://github.com/patriksimek/vm2

**Reply by DeanKamali** [reply]
@rijkvanzanten is this something you guys are working on? or it's a backlog item?

**Reply by claytongulick** [reply]
@DeanKamali we just sponsored this issue, so I hope so! 😄

**Reply by DeanKamali** [reply]
Thank you, I’m ready to chip in, but I don’t know how. 
looking forward to it!

**Comment by rijkvanzanten** [comment]
I'm wondering if we could somehow use something like https://github.com/geelen/npx-import to safely support external node modules as well. That'd be _huge_ for this operation 🤔

**Reply by claytongulick** [reply]
Wow, that would be amazing!

**Reply by Attacler** [reply]
> I'm wondering if we could somehow use something like https://github.com/geelen/npx-import to safely support external node modules as well. That'd be _huge_ for this operation 🤔

That would be huge but im wondering if that doesnt compromise safety? (like exposing process, access to the root filesystem/cli or something)

**Reply by rijkvanzanten** [reply]
Hmm not sure! Third party javascript is still javascript, so any nefarious thing a third party library can do can also be done by the user just copying in some bad JS. That's also why we're gonna run the user javascript in something like `vm2` to run the untrusted code in a sandbox. You would already be able to use the node modules that the API uses, however the main reason why I mentioned earlier that external node modules would be difficult, is that we don't have a way to "install" them securely yet. That library linked above could be the missing piece there

**Reply by rijkvanzanten** [reply]
vm2 allows a white list of allowed modules (https://www.npmjs.com/package/vm2), so that combined with npx-import could _theoretically_ allow for a configuration on the operation to set what modules you allow to run, and then have npx-import auto install them when they're actually used. That way we get the flexibility, but still ensure that you don't just run any untrusted code that messes up your computer 🤔

**Reply by DeanKamali** [reply]
Dose `npx-import` install them every time the Flow runs? Or only once during initial flow run?

**Reply by rijkvanzanten** [reply]
> Dose npx-import install them every time the Flow runs? Or only once during initial flow run?

From what I can tell, it installs it once, and relies on npm to return the globally temporary installed version on every subsequent call: https://github.com/geelen/npx-import#-doesnt-this-mean-dependencies-gets-repeatedly-downloaded--installed

**Reply by DeanKamali** [reply]
That’s good to hear! I was thinking it might nuke your cloud offering 😅

Exciting stuff!

**Reply by rijkvanzanten** [reply]
> I was thinking it might nuke your cloud offering 😅

Appreciate the concern 🙂 We're running the cloud instances in isolated containers for added security regardless, so the worst somebody could do is nuke their own project 😁

**Reply by Attacler** [reply]
> vm2 allows a white list of allowed modules (https://www.npmjs.com/package/vm2), so that combined with npx-import could _theoretically_ allow for a configuration on the operation to set what modules you allow to run, and then have npx-import auto install them when they're actually used. That way we get the flexibility, but still ensure that you don't just run any untrusted code that messes up your computer 🤔

Guess we would have to make a list of what it should or should not be able to do first?

**Reply by rijkvanzanten** [reply]
> Guess we would have to make a list of what it should or should not be able to do first?

The list is fairly short I think. It has to:

- Run an arbitrary (async) function that gets the current flow execution data as it's parameter
- Return a JSON object that can be used in subsequent operations

It can not:

- Access current process
- Access environment variables

Potential security no-no's[^1]:

- Access the network
- Access the filesystem

Loading third party modules _could_ be done safely, assuming that network/FS access is blocked in the sandbox that it runs in. However, we do need to make sure that libraries used are explicitly allowlisted (maybe as an ENV var) to prevent attack vectors in postinstall scripts and the like

[^1]: Security here is interesting. Technically speaking, you need to be an admin user to enter the JS that's ran. That being said, we are running "untrusted" JavaScript that's stored in the database, so we have to treat this as a "stored XSS" attack type of scenario.

**Reply by CHSchuepfer** [reply]
The Idea I came up with when I implemented my, janky, version of it was a way to snapshot the scope of an endpoint (graphql/system schema for example) and serve that as kind of a boilerplate for the script environment. I didn't come far bc I was a one man show and are only moderately firm in graphql schema. But if it can be implemented that way then you are absolutely safe bc the endpoint which acts as a source for the snapshot has all the security rails already implemented.


The negatives I came up with, for that idea are:

- you have to choose the right endpoint as source (not to little bc it limits the functionality in the end, too much and you open yourself up the security concerns)
- You have to keep an eye on the life circle of the script runtime
- The initial development work is kind of huge

The Pro's (imo) are:

- You don't have to rethink every edge case bc much of the things that are implemented in terms of checks and balances are already there and you have, for ongoing development, one point of contact which handles security (underlying scope of schema)
- You can even use nested scripts bc the boilerplate snapshot for the nested script can only access what the parent script has
- You can assign QOS like parameters more easily bc you can strip out informations more easily by blacklisting certain functions (FS Access for example) via query-schema



Just my 2 cents I wanted to throw in as I saw this discussion developing :)


so long

**Reply by claytongulick** [reply]
> Access the network

I get the reasoning behind this, especially with the cloud product - bad actors could write spambot or DDOS flows on free accounts and/or burn metered bandwith and blow out Directus' hosting costs.

It would be pretty sad to not be able to use fetch though if it were hosted locally - I wonder if network access could be made an ENV option?

**Reply by rijkvanzanten** [reply]
Oh the two under "no-no's" are meant as "These need some extra attention before opening up the floodgates" 🙂 I think allowing those to be enabled through an env var makes total sense. That will be safe for cloud too, as we enable it for paying customers (where we have decent accountability), but limit it on the free tiers 👍🏻

**Reply by manuelwehrli** [reply]
Are the reasons for limiting and sandboxing things only because of the cloud product? What about giving full access to everything in the self-hosted version? I mean, in a custom hook we can also use full js and external libraries. Why should this be different inside flows, which are only editable by an admin anyway?

**Reply by benhaynes** [reply]
Well, even outside of Cloud, we strive to ensure the platform is as safe and secure as possible. Keep in mind that some of our on-prem projects have "super-admins" that install and set up infra/config... then the platform admins have more limited access and shouldn't be given carte blanche control over scripting.

In any event, we actually figured out a secure way to handle this operation in a VM and it's in the process of being wrapped up... hahah, so it's win-win-win! 👍

**Reply by rijkvanzanten** [reply]
> Are the reasons for limiting and sandboxing things only because of the cloud product?

Actually no. The reason is security. Yes, that security obviously also applies to the cloud product, but is as important for self-hosted.

> What about giving full access to everything in the self-hosted version?

As with everything else, that will in fact be the case :) 

> I mean, in a custom hook we can also use full js and external libraries. Why should this be different inside flows, which are only editable by an admin anyway?

Cause admin users in Directus are not necessarily guaranteed to also be the server admins, or have access to the deployment environment, and because this code is being executed from the database, which means that there's an additional insecure entrance for untrusted code, as we don't necessarily know for a fact where that code came from (could be that the user entered it in the admin, submitted it through the API, entered directly in the database, or might even be a third party tool modifying the database). Therefore, we have to treat the code running under a flow operation as untrusted and  insecure by default

**Reply by manuelwehrli** [reply]
All valid points, I did not think about the use case of a server admin having more privileges than a directus admin in the self-hosted world. Might the sandboxing slow down the api performance at scale? Will there be a config option to enable untrusted execution and using full js beyond the possibilities inside the sandbox?

**Reply by rijkvanzanten** [reply]
> Might the sandboxing slow down the api performance at scale?

The sandboxing we're discussing here only applies to the "run arbitrary script" operation as part of a flow, nothing else in the API, so it shouldn't affect any of the currently existing performance. As for the performance of the sandbox itself, there will be some delay in spinning up vm in v8, but other than that the execution happens in the same process on the same machine, so the execution performance of the code itself should be about the same.

> Will there be a config option to enable untrusted execution and using full js beyond the possibilities inside the sandbox?

For this particular flow operation, there will be an environment variable that allows you to white list what node modules are allowed to be used, which in turn allows you to open up whatever use cases you want in the flow operation. Other than that, you already have "full js" inside the sandbox by default

**Comment by imiskolee** [comment]
I have create a custom branch to support eval on template parse stage, something looks like 

```json
{
   "field" : "{{  Eval(1+1)  }}"
}
```
